test/88_SIGNALduino_TOOL.pm

```diff
@@ -0,0 +1,1181 @@
+					$text2 =~ s/P$SP=//g;
+					$SyncPulse += $text2;
+				}
+
+				my $tol_min = abs($a[0]*(1-$tol));
+				my $tol_max = abs($a[0]*(1+$tol));
+
+				if (abs($text2) > $tol_min && abs($text2) < $tol_max) {
+					push(@Zeilen,$_);							# Zeile in array
+					$founded++;
+				}
+			}
+			$linecount++;
+		}
+		close InputFile;
+
+		readingsSingleUpdate($hash, "line_read" , $linecount, 0);
+		readingsSingleUpdate($hash, "state" , substr($cmd,4)." NOT in tol found!", 0) if ($founded == 0);
+		readingsSingleUpdate($hash, "state" , substr($cmd,4)." in tol found ($founded)", 0) if ($founded != 0);
+
+		return "ERROR: Your Attributes Filename_export is not definded!" if ($Filename_export eq "");
+		open(OutFile, ">$Filename_export");
+		for (@Zeilen) {
+			print OutFile $_."\n";
+		}
+		close OutFile;
+
+		return "ERROR: no $cmd with value $a[0] in tol!" if ($founded == 0);
+		return substr($cmd,4)." in tol found!";
+	}
+	
+	if ($cmd eq "invert_bitMsg" || $cmd eq "invert_hexMsg") {
+		return "ERROR: Your input failed!" if (not defined $a[0]);
+		return "ERROR: wrong value $a[0]! only [0-1]!" if ($cmd eq "invert_bitMsg" && not $a[0] =~ /^[0-1]+$/);
+		return "ERROR: wrong value $a[0]! only [a-fA-f0-9]!" if ($cmd eq "invert_hexMsg" && not $a[0] =~ /^[a-fA-f0-9]+$/);
+
+		if ($cmd eq "invert_bitMsg") {
+			$value = $a[0];
+			$value =~ tr/01/10/;															# ersetze ; durch ;;
+		} elsif ($cmd eq "invert_hexMsg") {
+			my $hlen = length($a[0]);
+			my $blen = $hlen * 4;
+			my $bitData = unpack("B$blen", pack("H$hlen", $a[0]));
+			$bitData =~ tr/01/10/;
+			$value = sprintf("%X", oct("0b$bitData"));		
+		}
+
+		return "Your $cmd is ready.\n\n  Input: $a[0]\n Output: $value";
+	}
+	
+	if ($cmd eq "change_hexMsg_to_bitMsg" || $cmd eq "change_bitMsg_to_hexMsg") {
+		return "ERROR: Your input failed!" if (not defined $a[0]);
+		return "ERROR: wrong value $a[0]! only [0-1]!" if ($cmd eq "change_bitMsg_to_hexMsg" && not $a[0] =~ /^[0-1]+$/);
+		return "ERROR: wrong value $a[0]! only [a-fA-f0-9]!" if ($cmd eq "change_hexMsg_to_bitMsg" && not $a[0] =~ /^[a-fA-f0-9]+$/);
+
+		if ($cmd eq "change_bitMsg_to_hexMsg") {
+			$value = sprintf("%x", oct( "0b$a[0]" ) );
+			$value = sprintf("%X", oct( "0b$a[0]" ) );
+			return "Your $cmd is ready.\n\nInput: $a[0]\n  Hex: $value";
+		} elsif ($cmd eq "change_hexMsg_to_bitMsg") {
+			$value = sprintf( "%b", hex( $a[0] ) );
+			return "Your $cmd is ready.\n\nInput: $a[0]\n  Bin: $value";
+		}
+	}
+
+	if ($cmd eq "FilterFile_doublePulse") {
+		return "ERROR: Your Attributes Filename_input is not definded!" if ($Filename_input eq "");
+
+		my $counterror = 0;
+		my $MUerror = 0;
+		my $MSerror = 0;
+		
+		open (InputFile,"<$path$Filename_input") || return "ERROR: No file ($Filename_input) found in $path directory from FHEM!";
+		while (<InputFile>){
+			if ($_ =~ /READredu:\sM(U|S);/s){
+				chomp ($_);										# Zeilenende entfernen
+				my $checkData = $_;
+
+				$_ =~ s/^.*;D=//g;						# cut bis D=
```

Comment:
Du weisst, dass Du hier die /g Option nicht benötigst und du anstelle vom Löschen auch einfach mit einer capture group an die Daten kommst?

https://regex101.com/r/qOLt95/1/

Reply:
Bei solchen Fällen lerne ich erst noch hinzu :) Es ging ja in erster Linie erstmal an die Umsetzung. Mit den Hinweisen werde ich nun jede Verbesserung gern übernehmen.

Ich bin schon so happy, das ich mich soweit „formen“ konnte was Perl im allgemeinen angeht hihihi.

Reply:
Die Dateien können gern in einen anderen Ordner. Es gibt dafür das Attribut Path was man dafür setzen kann.

Reply:
Ich meinte die Dateien des Tools :)

Reply:
@sidey79 ich habe mal noch die Anpassung des Verzeichnis vorgenommen und auch den Fix mit dem Pfad vorgenommen. Nun klappt es.

Die gesammelnten Daten in den ...Dispatch Dateien werden ab und zu mal aktualisiert und diese könntest du auch für diverse Tests mit nutzen. So haben wir was unsere Module angeht auch mal noch vereinzelte Nachrichten zum testen oder auch reproduzieren.